File: Hard/Word Ladder I/word-ladder-i.java
//{ Driver Code Starts
import java.util.*;
import java.lang.*;
import java.io.*;
class GFG
{
    public static void main(String[] args) throws IOException
    {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine().trim());
        while(T-->0)
        {
            int n = Integer.parseInt(br.readLine().trim());
            String[] wordList = new String[n];
            for(int i = 0; i < n; i++){
                wordList[i] = br.readLine().trim();
            }
            String startWord, targetWord;
            startWord = br.readLine().trim();
            targetWord = br.readLine().trim();
            Solution obj = new Solution();
            int ans = obj.wordLadderLength(startWord, targetWord, wordList);
            System.out.println(ans);
       }
    }
}

// } Driver Code Ends

class pair
{
    String str;
    int level;
    pair(String s,int x)
    {
        str=s;
        level=x;
    }
}
class Solution
{
    public void helper(String s,int level,HashSet<String> set,Queue<pair> q)
    {
        char ch[]=s.toCharArray();
        int i,j,n=ch.length;
        for(i=0;i<n;i++)
        {
            char temp[]=s.toCharArray();
            for(j=0;j<26;j++)
            {
                temp[i]=(char)(j+'a');
                if(set.contains(new String(temp)))
                {
                    q.offer(new pair(new String(temp),level+1));
                    set.remove(new String(temp));
                }
            }
        }
    }
    public int wordLadderLength(String startWord, String targetWord, String[] wordList)
    {
        // Code here
        int mini=Integer.MAX_VALUE;
        Queue<pair> q=new LinkedList<>();
        q.offer(new pair(startWord,1));
        HashSet<String> set=new HashSet<>();
        for(String x:wordList)
        set.add(x);
        set.remove(startWord);
        while(!q.isEmpty())
        {
            pair top=q.poll();
            if(top.str.equals(targetWord))
            {
                return top.level;
            }
            helper(top.str,top.level,set,q);
        }
        return mini==Integer.MAX_VALUE?0:mini;
    }
}

File: Hard/Word Ladder I/word-ladder-i.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends



#include <iostream>
#include <vector>
#include <queue>
#include <unordered_set>

using namespace std;

class Pair {
public:
    string str;
    int level;
    
    Pair(string s, int x) : str(s), level(x) {}
};

class Solution {
public:
    void helper(string s, int level, unordered_set<string>& wordSet, queue<Pair>& q) {
        int n = s.length();
        for (int i = 0; i < n; i++) {
            string temp = s;
            for (char j = 'a'; j <= 'z'; j++) {
                temp[i] = j;
                if (wordSet.find(temp) != wordSet.end()) {
                    q.push(Pair(temp, level + 1));
                    wordSet.erase(temp);
                }
            }
        }
    }

    int wordLadderLength(string beginWord, string endWord, vector<string>& wordList) {
        int mini = INT_MAX;
        queue<Pair> q;
        q.push(Pair(beginWord, 1));
        unordered_set<string> wordSet(wordList.begin(), wordList.end());
        wordSet.erase(beginWord);
        
        while (!q.empty()) {
            Pair top = q.front();
            q.pop();
            
            if (top.str == endWord) {
                return top.level;
            }
            
            helper(top.str, top.level, wordSet, q);
        }
        
        return mini == INT_MAX ? 0 : mini;
    }
};


//{ Driver Code Starts.
int main(){
	int tc;
	cin >> tc;
	while(tc--){
		int n;
		cin >> n;
		vector<string>wordList(n);
		for(int i = 0; i < n; i++)cin >> wordList[i];
		string startWord, targetWord;
		cin >> startWord >> targetWord;
		Solution obj;
		int ans = obj.wordLadderLength(startWord, targetWord, wordList);
		cout << ans << "\n";
	}
	return 0;
}
// } Driver Code Ends